src/usr.bin/tr/tr.1
.\"	$OpenBSD: tr.1,v 1.9 2003/06/10 09:12:12 jmc Exp $
.\"	$NetBSD: tr.1,v 1.5 1994/12/07 08:35:13 jtc Exp $
.\"
.\" Copyright (c) 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" the Institute of Electrical and Electronics Engineers, Inc.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)tr.1	8.1 (Berkeley) 6/6/93
.\"
.Dd $Mdocdate$
.Dt TR 1
.Os
.Sh NAME
.Nm tr
.Nd translate characters
.Sh SYNOPSIS
.Nm tr
.Op Fl cs
.Ar string1 string2
.Nm tr
.Op Fl c
.Fl d
.Ar string1
.Nm tr
.Op Fl c
.Fl s
.Ar string1
.Nm tr
.Op Fl c
.Fl ds
.Ar string1 string2
.Sh DESCRIPTION
The
.Nm
utility copies the standard input to the standard output with substitution
or deletion of selected characters.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl c
Complements the set of characters in
.Ar string1 ;
for instance,
.Dq Fl c No ab
includes every character except for
.Dq a
and
.Dq b .
.It Fl d
The
.Fl d
option causes characters to be deleted from the input.
.It Fl s
The
.Fl s
option squeezes multiple occurrences of the characters listed in the last
operand (either
.Ar string1
or
.Ar string2 )
in the input into a single instance of the character.
This occurs after all deletion and translation is completed.
.El
.Pp
In the first synopsis form, the characters in
.Ar string1
are translated into the characters in
.Ar string2
where the first character in
.Ar string1
is translated into the first character in
.Ar string2
and so on.
If
.Ar string1
is longer than
.Ar string2 ,
the last character found in
.Ar string2
is duplicated until
.Ar string1
is exhausted.
.Pp
In the second synopsis form, the characters in
.Ar string1
are deleted from the input.
.Pp
In the third synopsis form, the characters in
.Ar string1
are compressed as described for the
.Fl s
option.
.Pp
In the fourth synopsis form, the characters in
.Ar string1
are deleted from the input, and the characters in
.Ar string2
are compressed as described for the
.Fl s
option.
.Pp
The following conventions can be used in
.Ar string1
and
.Ar string2
to specify sets of characters:
.Bl -tag -width [:equiv:]
.It character
Any character not described by one of the following conventions
represents itself.
.It \eoctal
A backslash followed by 1, 2, or 3 octal digits represents a character
with that encoded value.
To follow an octal sequence with a digit as a character, left zero-pad
the octal sequence to the full 3 octal digits.
.It \echaracter
A backslash followed by certain special characters maps to special
values.
.Pp
.Bl -column
.It \ea	<alert character>
.It \eb	<backspace>
.It \ef	<form-feed>
.It \en	<newline>
.It \er	<carriage return>
.It \et	<tab>
.It \ev	<vertical tab>
.El
.Pp
A backslash followed by any other character maps to that character.
.It c-c
Represents the range of characters between the range endpoints, inclusively.
.It [:class:]
Represents all characters belonging to the defined character class.
Class names are:
.Pp
.Bl -column
.It alnum	<alphanumeric characters>
.It alpha	<alphabetic characters>
.It blank	<blank characters>
.It cntrl	<control characters>
.It digit	<numeric characters>
.It graph	<graphic characters>
.It lower	<lower-case alphabetic characters>
.It print	<printable characters>
.It punct	<punctuation characters>
.It space	<space characters>
.It upper	<upper-case characters>
.It xdigit	<hexadecimal characters>
.El
.Pp
Characters
in the classes are in unspecified order.
In this implementation, they are usually in Unicode
ascending order.
.Pp
For specific information as to which ASCII or multibyte characters are included
in these classes, see
.Xr wctype 3
and related manual pages.
.It [=equiv=]
Represents all characters or collating (sorting) elements belonging to
the same equivalence class as
.Ar equiv .
If
there is a secondary ordering within the equivalence class, the characters
are ordered in ascending sequence.
Otherwise, they are ordered after their encoded values.
An example of an equivalence class might be
.Dq c
and
.Dq ch
in Spanish;
English has no equivalence classes.
.It [#*n]
Represents
.Ar n
repeated occurrences of the character represented by
.Ar # .
This
expression is only valid when it occurs in
.Ar string2 .
If
.Ar n
is omitted or is zero, it is be interpreted as large enough to extend
.Ar string2
sequence to the length of
.Ar string1 .
If
.Ar n
has a leading zero, it is interpreted as an octal value; otherwise,
it's interpreted as a decimal value.
.El
.Pp
The
.Nm
utility exits 0 on success or >0 if an error occurred.
.Sh EXAMPLES
The following examples are shown as given to the shell:
.Pp
Create a list of the words in file1, one per line, where a word is taken to
be a maximal string of letters.
.Pp
.D1 Li "$ tr -cs \*q[:alpha:]\*q \*q\en\*q < file1"
.Pp
Translate the contents of file1 to upper-case.
.Pp
.D1 Li "$ tr \*q[:lower:]\*q \*q[:upper:]\*q < file1"
.Pp
Strip out non-printable characters from file1.
.Pp
.D1 Li "$ tr -cd \*q[:print:]\*q < file1"
.Sh SEE ALSO
.Xr sed 1
.Sh STANDARDS
System V has historically implemented character ranges using the syntax
.Dq [c-c]
instead of the
.Dq c-c
used by historic BSD implementations and
standardised by POSIX.
System V shell scripts should work under this implementation as long as
the range is intended to map in another range, i.e., the command
.Dq tr\ [a-z]\ [A-Z]
will work as it will map the
.Dq [
character in
.Ar string1
to the
.Dq [
character in
.Ar string2 .
However, if the shell script is deleting or squeezing characters as in
the command
.Dq tr\ -d\ [a-z] ,
the characters
.Dq [
and
.Dq \]
will be
included in the deletion or compression list, which would not have happened
under an historic System V implementation.
Additionally, any scripts that depended on the sequence
.Dq a-z
to represent the three characters
.Dq a ,
.Dq - ,
and
.Dq z
will have to be rewritten as
.Dq a\e-z .
.Pp
The
.Nm
utility has historically not permitted the manipulation of NUL bytes in
its input and, additionally, has stripped NUL's from its input stream.
This implementation has removed this behavior as a bug.
.Pp
The
.Nm
utility has historically been extremely forgiving of syntax errors:
for example, the
.Fl c
and
.Fl s
options were ignored unless two strings were specified.
This implementation will not permit illegal syntax.
.Pp
The
.Nm
utility is expected to be
.St -p1003.2
compatible.
It should be noted that the feature wherein the last character of
.Ar string2
is duplicated if
.Ar string2
has less characters than
.Ar string1
is permitted by POSIX but is not required.
Shell scripts attempting to be portable to other POSIX systems should use
the
.Dq [#*]
convention instead of relying on this behavior.


src/usr.bin/tr/tr.c
/*	$OpenBSD: tr.c,v 1.13 2004/09/15 22:12:19 deraadt Exp $	*/
/*	$NetBSD: tr.c,v 1.5 1995/08/31 22:13:48 jtc Exp $	*/

/*
 * Copyright (c) 2007, 2008, 2009
 *	Thorsten Glaser <tg@mirbsd.org>
 * Copyright (c) 1988, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/types.h>

#include <unistd.h>
#include <err.h>
#include <errno.h>
#include <stdbool.h>
#include <wctype.h>

#include "extern.h"

__COPYRIGHT("@(#) Copyright (c) 1988, 1993\n\
	The Regents of the University of California.  All rights reserved.\n");
__SCCSID("@(#)tr.c	8.2 (Berkeley) 5/4/95");
__RCSID("$MirOS: src/usr.bin/tr/tr.c,v 1.5 2009/09/25 23:02:08 tg Exp $");

static wchar_t string1[NCHARS], string2[NCHARS];

STR s1 = {
	.which = STRING1,
	.state = NORMAL,
	.cnt = 0,
	.lastch = OOBCH,
	.equiv = { 0, OOBCH },
	.set = NULL,
	.use_wctrans = false,
	.str = NULL
};
STR s2 = {
	.which = STRING2,
	.state = NORMAL,
	.cnt = 0,
	.lastch = OOBCH,
	.equiv = { 0, OOBCH },
	.set = NULL,
	.use_wctrans = false,
	.str = NULL
};

static void setup(wchar_t *, char *, STR *, bool);
static void usage(void) __dead;

#define getwcf() __extension__({				\
	wint_t getwcf_c;					\
								\
	if ((getwcf_c = getwchar()) == WEOF && ferror(stdin) &&	\
	    errno == EILSEQ) {					\
		clearerr(stdin);				\
		getwcf_c = 0xFFFD;				\
	}							\
	(getwcf_c);						\
})

int
main(int argc, char *argv[])
{
	wint_t ch;
	wchar_t cnt, lastch, *p;
	int ich;
	bool cflag, dflag, sflag, isstring2;

#ifndef __MirBSD__
	setlocale(LC_ALL, "");
#endif

	for (lastch = 0; lastch < NCHARS; ++lastch)
		string1[lastch] = lastch;

	cflag = dflag = sflag = false;
	while ((ich = getopt(argc, argv, "cds")) != -1)
		switch (ich) {
		case 'c':
			cflag = true;
			break;
		case 'd':
			dflag = true;
			break;
		case 's':
			sflag = true;
			break;
		case '?':
		default:
			usage();
		}
	argc -= optind;
	argv += optind;

	switch(argc) {
	case 0:
	default:
		usage();
		/* NOTREACHED */
	case 1:
		isstring2 = false;
		break;
	case 2:
		isstring2 = true;
		break;
	}

	/*
	 * tr -ds [-c] string1 string2
	 * Delete all characters (or complemented characters) in string1.
	 * Squeeze all characters in string2.
	 */
	if (dflag && sflag) {
		if (!isstring2)
			usage();

		setup(string1, argv[0], &s1, cflag);
		setup(string2, argv[1], &s2, false);

		for (lastch = OOBCH; (ch = getwcf()) != WEOF; )
			if (!string1[ch] && (!string2[ch] || lastch != ch)) {
				lastch = ch;
				putwchar(ch);
			}
		exit(0);
	}

	/*
	 * tr -d [-c] string1
	 * Delete all characters (or complemented characters) in string1.
	 */
	if (dflag) {
		if (isstring2)
			usage();

		setup(string1, argv[0], &s1, cflag);

		while ((ch = getwcf()) != WEOF)
			if (!string1[ch])
				putwchar(ch);
		exit(0);
	}

	/*
	 * tr -s [-c] string1
	 * Squeeze all characters (or complemented characters) in string1.
	 */
	if (sflag && !isstring2) {
		setup(string1, argv[0], &s1, cflag);

		for (lastch = OOBCH; (ch = getwcf()) != WEOF; )
			if (!string1[ch] || lastch != ch) {
				lastch = ch;
				putwchar(ch);
			}
		exit(0);
	}

	/*
	 * tr [-cs] string1 string2
	 * Replace all characters (or complemented characters) in string1 with
	 * the character in the same position in string2.  If the -s option is
	 * specified, squeeze all the characters in string2.
	 */
	if (!isstring2)
		usage();

	s1.str = ambstowcs(argv[0]);
	s2.str = ambstowcs(argv[1]);
	if ((strstr(argv[0], "[:lower:]") && strstr(argv[1], "[:upper:]")) ||
	    (strstr(argv[1], "[:lower:]") && strstr(argv[0], "[:upper:]"))) {
		s1.use_wctrans = true;
		s2.use_wctrans = true;
	}

	if (cflag)
		for (cnt = NCHARS, p = string1; cnt--;)
			*p++ = OOBCH;

	if (!next(&s2))
		errx(1, "empty string2");

	/* If string2 runs out of characters, use the last one specified. */
	ch = s2.lastch;
	if (sflag)
		while (next(&s1)) {
			string1[s1.lastch] = ch = s2.lastch;
			string2[ch] = 1;
			(void)next(&s2);
		}
	else
		while (next(&s1)) {
			string1[s1.lastch] = ch = s2.lastch;
			(void)next(&s2);
		}

	if (cflag)
		for (cnt = 0, p = string1; cnt < NCHARS; ++p, ++cnt)
			*p = *p == OOBCH ? ch : cnt;

	if (sflag)
		for (lastch = OOBCH; (ch = getwcf()) != WEOF;) {
			ch = string1[ch];
			if (!string2[ch] || lastch != ch) {
				lastch = ch;
				putwchar(ch);
			}
		}
	else
		while ((ch = getwcf()) != WEOF)
			putwchar(string1[ch]);
	exit(0);
}

static void
setup(wchar_t *string, char *arg, STR *str, bool cflag)
{
	wchar_t cnt, *p;

	str->str = ambstowcs(arg);
	bzero(string, NCHARS * sizeof (wchar_t));
	while (next(str))
		string[str->lastch] = 1;
	if (cflag)
		for (p = string, cnt = NCHARS; cnt--; ++p)
			*p = !*p;
}

static void
usage(void)
{
	fprintf(stderr, "usage:\ttr [-cs] string1 string2\n"
	    "\ttr [-c] -d string1\n"
	    "\ttr [-c] -s string1\n"
	    "\ttr [-c] -ds string1 string2\n");
	exit(1);
}
